plugin/src/helpers/config.ts

```diff
@@ -108,3 +123,82 @@ export const configureHandlerFunctions = ({ netlifyConfig, publish, ignore = []
+  const { path, headers } = buildHeaderParams
+
+  return {
+    for: path,
+    values: headers.reduce((builtHeaders, { key, value }) => {
+      builtHeaders[key] = value
+
+      return builtHeaders
+    }, {}),
+  }
+}
+
+// Replace the pattern :path* at the end of a path with * since it's a named splat which the Netlify
+// configuration does not support.
+const sanitizePath = (path: string) => path.replace(/:[^*/]+\*$/, '*')
+
+/**
+ * Persist NEXT.js custom headers to the Netlify configuration so the headers work with static files
+ * See {@link https://nextjs.org/docs/api-reference/next.config.js/headers} for more information on custom
+ * headers in Next.js
+ *
+ * @param nextConfig - The NextJS configuration
+ * @param netlifyHeaders - Existing headers that are already configured in the Netlify configuration
+ */
+export const generateCustomHeaders = (nextConfig: NextConfig, netlifyHeaders: NetlifyHeaders = []) => {
+  // The routesManifest is the contents of the routes-manifest.json file which will already contain the generated
+  // header paths which take locales and base path into account since this runs after the build. The routes-manifest.json
+  // file is located at demos/default/.next/routes-manifest.json once you've build the demo site.
+  const {
+    routesManifest: { headers: customHeaders = [] },
+    i18n,
+  } = nextConfig
+
+  // Skip `has` based custom headers as they have more complex dynamic conditional header logic
+  // that currently isn't supported by the Netlify configuration.
+  // Also, this type of dynamic header logic is most likely not for SSG pages.
+  for (const { source, headers, locale: localeEnabled } of customHeaders.filter((customHeader) => !customHeader.has)) {
+    // Explicitly checkling false to make the check simpler.
+    // Locale specific paths are excluded only if localeEnabled is false. There is no true value for localeEnabled. It's either
+    // false or undefined, where undefined means it's true.
+    //
+    // Again, the routesManifest has already been generated taking locales into account, but the check is required
+    // so  the paths can be properly set in the Netlify configuration.
+    const useLocale = i18n?.locales?.length > 0 && localeEnabled !== false
+
+    if (useLocale) {
+      const { locales } = i18n
+      const joinedLocales = locales.join('|')
+
+      /**
+       *  converts e.g.
+       *  /:nextInternalLocale(en|fr)/some-path
+       *  to a path for each locale
+       *  /en/some-path and /fr/some-path as well as /some-path (default locale)
+       */
+      const defaultLocalePath = sanitizePath(source).replace(`/:nextInternalLocale(${joinedLocales})`, '')
+
+      netlifyHeaders.push(buildHeader({ path: defaultLocalePath, headers }))
+
+      for (const locale of locales) {
+        const path = sanitizePath(source).replace(`:nextInternalLocale(${joinedLocales})`, locale)
+
+        netlifyHeaders.push(buildHeader({ path, headers }))
+      }
```

Comment:
[sand] It's small enough that it's not something that necessarily needs to be addressed now, but this code looks like it could be broken into its own helper function because it deals entirely with local specific path when that functionality is enabled.

Reply:
I'm going to hold off on this for now.